Source/Core/RequestScheduler.js

```diff
@@ -0,0 +1,383 @@
+        RequestState,
+        Uri,
+        when) {
+    'use strict';
+
+    /**
+     * Tracks the number of active requests and prioritizes incoming requests.
+     *
+     * @exports RequestScheduler
+     *
+     * @private
+     */
+    function RequestScheduler() {
+    }
+
+    /**
+     * The maximum number of simultaneous active requests. Un-throttled requests do not observe this limit.
+     * @type {Number}
+     * @default 50
+     */
+    RequestScheduler.maximumRequests = 50;
+
+    /**
+     * The maximum number of simultaneous active requests per server. Un-throttled requests do not observe this limit.
+     * @type {Number}
+     * @default 50
+     */
+    RequestScheduler.maximumRequestsPerServer = 6;
+
+    /**
+     * The maximum size of the priority heap. This limits the number of requests that are sorted by priority. Only applies to requests that are not yet active.
+     * @type {Number}
+     * @default 20
+     */
+    RequestScheduler.priorityHeapSize = 20;
+
+    /**
+     * Specifies if the request scheduler should throttle incoming requests, or let the browser queue requests under its control.
+     * @type {Boolean}
+     * @default true
+     */
+    RequestScheduler.debugThrottle = true;
+
+    /**
+     * When true, log statistics to the console every frame
+     * @type {Boolean}
+     * @default false
+     */
+    RequestScheduler.debugShowStatistics = false;
+
+    defineProperties(RequestScheduler, {
+        /**
+         * Returns the statistics used by the request scheduler.
+         *
+         * @memberof RequestScheduler
+         *
+         * @type Object
+         * @readonly
+         */
+        statistics : {
+            get : function() {
+                return statistics;
+            }
+        }
+    });
+
+    function sortRequests(a, b) {
+        return a.distance - b.distance;
+    }
+
+    var statistics = {
+        numberOfAttemptedRequests : 0,
+        numberOfActiveRequests : 0,
+        numberOfCancelledRequests : 0,
+        numberOfCancelledActiveRequests : 0,
+        numberOfFailedRequests : 0,
+        numberOfActiveRequestsEver : 0
+    };
+
+    var requestHeap = new Heap(sortRequests);
+    requestHeap.maximumLength = RequestScheduler.priorityHeapSize;
+    requestHeap.reserve(RequestScheduler.priorityHeapSize);
+    var activeRequests = [];
+
+    var numberOfActiveRequestsByServer = {};
+
+    var pageUri = typeof document !== 'undefined' ? new Uri(document.location.href) : new Uri();
+
+    /**
+     * Get the server name from a given url.
+     *
+     * @param {String} url The url.
+     * @returns {String} The server name.
+     */
+    RequestScheduler.getServer = function(url) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string('url', url);
+        //>>includeEnd('debug');
+
+        var uri = new Uri(url).resolve(pageUri);
+        uri.normalize();
+        var serverName = uri.authority;
+        if (!/:/.test(serverName)) {
```

Comment:
Took me a minute to figure out what was going on here.  A comment would go a long way.